pytests/test_debugging.py

```diff
@@ -0,0 +1,504 @@
+        self.pause = None
+        return self.send_message(DebuggerClientMessage.DEBUGGER_RESUME)
+
+    """
+    Resumes the execution (if it was paused), waits until the next
+    pause happens and returns all the messages in between.
+    """
+    def resume_and_wait_until_pauses(self) -> [dict]:
+        messages = self.resume()
+        while True:
+            message = self.receive_event()
+            messages.append(message)
+            if "method" in message and message["method"] == "Debugger.paused":
+                self.pause = message
+                break
+        return messages
+
+    """
+    Debugger: step into the stack frame.
+    """
+    def step_into(self) -> [dict]:
+        return self.send_message(DebuggerClientMessage.DEBUGGER_STEP_INTO)
+
+    """
+    Debugger: step out of the current stack frame.
+    """
+    def step_out(self) -> [dict]:
+        return self.send_message(DebuggerClientMessage.DEBUGGER_STEP_OUT)
+
+    """
+    Debugger: step over the next instruction.
+    """
+    def step_over(self) -> [dict]:
+        return self.send_message(DebuggerClientMessage.DEBUGGER_STEP_OVER)
+
+
+def deploy_script(env: Env):
+    env.expect('TFUNCTION', 'DELETE', 'foo')
+    env.expect('TFUNCTION', 'LOAD', 'debug', SCRIPT).contains('The V8 remote debugging server is waiting for a connection')
+
+@gearsTest()
+def testDeployConnectStartDisconnect(env: Env):
+    deploy_script(env)
+    client = DebuggerClient(env)
+    client.disconnect()
+
+@gearsTest()
+def testSetBreakpointAndHitIt(env: Env):
+    deploy_script(env)
+    client = DebuggerClient(env)
+    breakpoint = client.set_breakpoint(2, 1)
+    env.assertEqual(len(client.breakpoints), 1)
+    # Now proceed from the very first instruction to the very first
+    # breakpoint we've just set.
+    client.resume_and_wait_until_pauses()
+    env.assertIsNotNone(client.pause)
+    env.assertContains("callFrames", client.pause["params"])
+    # Check that we hit the breakpoint we set above.
+    env.assertEqual(breakpoint.get_id(), client.pause["params"]["hitBreakpoints"][0])
+
+    client.disconnect()
+
+@gearsTest()
+def testSetBreakpointAndRemoveIt(env: Env):
+    deploy_script(env)
+    client = DebuggerClient(env)
+    breakpoint = client.set_breakpoint(2, 1)
+    env.assertEqual(len(client.breakpoints), 1)
+    breakpoint.remove()
+    # Now proceed from the very first instruction till the end.
+    env.assertEqual({}, client.resume()[0]["result"])
+    # The next message should be a "resumed" event from the server.
+    resumed_event = client.receive_event()
+    env.assertEqual("Debugger.resumed", resumed_event["method"])
+    # Now wait for the possible pause event to happen. It should not
+    # happen now, for we have no breakpoints set, and the script
+    # execution should just continue with no problem. We wait for
+    # five seconds to let *any* sort of execution to continue on any VM,
+    # so to avoid the flakiness of the test result.
```

Comment:
Maybe we should also do a simple `ping` command and verify that redis is responsive?

Reply:
Added a ping check.